### DISCUSSION TITLE
vfkit VM console

### DISCUSSION THREAD

**Discussion by mrclramos** [crc-org/crc#3335] [Q&A]
Hello

Is it possible to open the local/serial console for the vfkit VM on MacOS using the '_screen_' command or similar tool?

Thanks in advance.

**Comment by cfergeau** [comment]
This is not possible with the vfkit instance started by crc. The output of the serial console should be available in `~/.crc/machines/crc/vfkit.log`. And it's possible to get a ssh session to the VM, see https://github.com/code-ready/crc/wiki/Debugging-guide
Are you trying to debug some specific issue?

**Comment by mrclramos** [comment]
Hello @cfergeau,

I saw the serial output log and that is why I asked about direct access. The intend is to use the native Virtualization Framework and have other VMs interacting with the crc cluster.

Is there any documentation on using vfkit to start other VMs or should I take a look on Code-Hex/vz for that?

Thanks in advance.

**Reply by cfergeau** [reply]
Hm, yeah vfkit is thoroughly lacking documentation :-/ I should work on that! The best way to see what is supported at the moment is to get the `vfkit` command line crc uses :-/

```
Users/teuf/.crc/bin/vfkit --cpus 2 --memory 2048 --kernel /Users/teuf/.crc/cache/crc_podman_vfkit_4.1.1_amd64/vmlinuz-5.18.11-200.fc36.x86_64 --initrd /Users/teuf/.crc/cache/crc_podman_vfkit_4.1.1_amd64/initramfs-5.18.11-200.fc36.x86_64.img --kernel-cmdline 'console=hvc0 [snip]' --device virtio-serial,logFilePath=/Users/teuf/.crc/machines/crc/vfkit.log --device virtio-rng --device virtio-blk,path=/Users/teuf/.crc/machines/crc/crc.img --device virtio-vsock,port=1024,socketURL=/Users/teuf/.crc/tap.sock
```

It's already quite long, but there's https://github.com/code-ready/vfkit/blob/main/pkg/client/cmdline.go to help create the commandline through a golang API.

I'd like `vfkit` to be as generic as possible (ie expose everything that is bound in github.com/Code-Hex/vz), but for now it's indeed very CRC-specific, and I did not implement much besides what it needed.

Regarding serial output, `vfkit` only has support for `-d virtio-serial,logFilePath=...`, but it should be possible to add a way of doing what you need. I usually try to take inspiration from QEMU, so maybe `-d virtio-serial,stdio`, `-d virtio-serial, pty`. Or would you need another way of accessing the serial console?

Last but not least, should you go with `vfkit` or `Code-Hex/vz`? The answer is that it depends. The big limitation of using `vz` directly is that the VM it starts will only be available as long as the process which started it is alive. If you have a go application which will stay alive for much longer than your VM, then you can go with vz. If you want your VMs lifetime to be independent from the lifetime of the process which starts them, then you will need to wrap the `vz` API in a separate binary. Which is what `vfkit` is doing. In that case, I'd recommend reusing vfkit, and filing patches to add the features you are missing ^^

**Reply by mrclramos** [reply]
Thanks for the reply.

I will take a look on the cmdline.go and start from there.

**Reply by cfergeau** [reply]
I've started adding doc for the command line https://github.com/cfergeau/vfkit/blob/doc/doc/usage.md and for the go client code https://github.com/cfergeau/vfkit/blob/doc/pkg/client/cmdline.go
